### PR TITLE
Add config path flag to cli

### PR DESCRIPTION
**Pull request by lazreg87**
**Description**

Add flag to be able to specify path of config file.
Fixes issue #1115 and was also discussed in https://github.com/firebase/firebase-tools/issues/590#issuecomment-406604437

OBS:
You still need to have a firebase.json present to detect project root. Didn’t want to tackle nesting(or messing) that up in this PR

**Scenarios Tested**
- Added unit tests for Config.load method
- Manually tested `firebase init —config-path <path>` to see that change does not break existing command. Should maybe discuss if we copy content of supplied config to generated firebase.json
- Manually tested `firebase serve —config-path <path>` to see that change does not break existing command.
- Manually tested `firebase deploy —config-path <path>` using different config files, with different predeploy commands to see correct config was loaded.

**Sample Commands**

`firebase serve —config-path <path>`
`firebase deploy —config-path <path>`

### CODE REVIEW

File: src/command.js [hunk 1]
```diff
@@ -144,6 +144,11 @@ Command.prototype._prepare = function(options) {
     logger.transports.console.level = "none";
   }
 
+  var configPath = utils.getInheritedOption(options, "configPath");
+  if (configPath) {
```
Comment (bkendall) [inline src/command.js]: is this check necessary? if it never finds a value, it returns undefined, which is the same as not setting the value (below)

File: src/index.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ program.option("--token <token>", "supply an auth token for this command");
 program.option("--non-interactive", "error out of the command instead of waiting for prompts");
 program.option("--interactive", "force interactive shell treatment even when not detected");
 program.option("--debug", "print verbose debug output and keep a debug log file");
+program.option("--config-path <path>", "path to custom config file (default: Firebase.json)");
```
Comment (bkendall) [inline src/index.js]:
```suggestion
program.option("--config-path <path>", "path to custom config file (default: `firebase.json`)");
```

File: src/test/config.spec.js [hunk 3]
```diff
@@ -99,4 +99,20 @@ describe("Config", function() {
...
+    it("should try to load default config if no config flag is provided", function() {
+      var config = Config.load({ cwd: _fixtureDir("config-imports") });
+      expect(config._src.hosting).to.equal("hosting.json");
+      expect(config._src.rules).to.equal("rules.json");
+    });
+
+    it("should load config from specified path when config flag is provided", function() {
+      var config = Config.load({
+        cwd: _fixtureDir("valid-config"),
+        configPath: "../valid-config/firebase.json",
```
Comment (bkendall) [inline src/test/config.spec.js]: this is the default behavior, isn't it? If you removed all the code that was added, this test would still pass, I think. Should the `cwd` be changed to `invalid-config` and the `configPath` pointed at the correct `firebase.json` to correctly test this behavior (or point to a different file, that isn't `firebase.json`, but would still be valid?)

File: src/command.js (hunk 1 above)
Comment (lazreg87) [reply src/command.js]: You are right, unnecessary check. Removed it

File: src/index.js (hunk 2 above)
Comment (lazreg87) [reply src/index.js]: done

File: src/test/config.spec.js (hunk 3 above)
Comment (lazreg87) [reply src/test/config.spec.js]: You were correct, the test still passed on master. Fixed the test according to your comment

**Comment by samtstern** [comment]
@bkendall did this ever go through API council?  Should it?

**Comment by LiveDuo** [comment]
Is there any way I can npm install this instead of firebase-tools?

**Comment by lazreg87** [comment]
@bkendall, still pending approval by API council?

**Comment by rhodgkins** [comment]
Is there any movement on this?

**Comment by samtstern** [comment]
Hey everyone this got lost in the shuffle but I'm gonna sponsor this change through API council.  @lazreg87 once it's approved I'll let you know and if you're still interested in doing this PR I'll help you get through the conflicts.  Otherwise I'll implement it myself.

**Comment by samtstern** [comment]
I just sent out the internal API proposal, should have official word on this in a week or maybe two at most.

**Comment by samtstern** [comment]
@lazreg87 oh wow I feel bad!  By the time the API was approved and I got around to writing https://github.com/firebase/firebase-tools/pull/2467 I totally forgot about this PR.  I didn't mean to steal your thunder and I should have worked with you to get this one into shape instead of making my own.

Thank you for starting the conversation and making this API happen.  Let us know if the `--config` flag works as expected once it's released!

**Comment by lazreg87** [comment]
No worries @samtstern, I'm happy the flag got added. Thx

**Comment by baermathias** [comment]
@lazreg87 So the flag is already available? like this? 
`firebase deploy --config foo.firebase.json`

**Comment by samtstern** [comment]
@baermathias yep just update to the latest version and it's available!

**Pull request closed without merging**